torch/lib/c10d/reducer.cpp

```diff
@@ -526,7 +536,18 @@ void Reducer::autograd_hook(VariableIndex index) {
     // to mark it in local_used_maps_. During no_sync session, the same var can
     // be set multiple times, which is OK as does not affect correctness. As
     // long as it is used once during no_sync session, it is marked as used.
-    local_used_maps_[index.replica_index][index.variable_index] = 1;
+    // Only set it as locally used if the grad is defined. Otherwise, hooks
+    // could sometimes be triggered with undefined grads, and if this happens
+    // globally, we don't want to touch the .grad field of the param.
+
+    auto& variable = get_param_from_index(index);
+    runGradCallbackForVariable(variable, [&](auto& grad) {
```

Comment:
why the runGradCallbackForVariable is needed here? grad is not mutated anyway

Reply:
We have to access `variable.grad()` right? When running in just DDP, this can be done with `variable.grad()` directly, but I believe this callback is needed when running in RPC context, where the grad is not available in variable.grad() field.

Reply:
a, you are right!